### PR TITLE
Fix EDR when several bugs for same environment

### PR DESCRIPTION
**Pull request by enzofab91**
### **Jira cards:**
[EDR-221](https://rootstrap.atlassian.net/browse/EM-221) - EDR not taking updated bugs into account
[EDR-222](https://rootstrap.atlassian.net/browse/EM-222) - EDR showing empty string when environment is null

### **Description:**
This PR fixes a bug that is getting wrong quantity issues when are several bugs in the same period.
It also fixes the view when environment is `null`.

### CODE REVIEW

File: app/services/metrics/defect_escape_rate/per_project.rb [hunk 1]
```diff
@@ -27,9 +27,14 @@ def query(interval)
         ]]
       end
 
+      # :reek:DuplicateMethodCall
       def bugs_by_environment(bug_issues)
         bug_issues.each_with_object({}) do |bug, result|
```
Comment (horacio) [inline app/services/metrics/defect_escape_rate/per_project.rb:32]: Maybe we can set a default for these keys? Something like...

```suggestion
        bug_issues.each_with_object(Hash.new(1)) do |bug, result|
```

File: Gemfile.lock [hunk 2]
```diff
@@ -149,7 +149,7 @@ GEM
     faraday (1.1.0)
       multipart-post (>= 1.2, < 3)
       ruby2_keywords
-    ffi (1.11.3)
+    ffi (1.15.3)
```
Comment (santib) [inline Gemfile.lock:152]: in general, we shouldn't upgrade gems as part of other tickets

File: app/services/metrics/defect_escape_rate/per_project.rb [hunk 3]
```diff
@@ -27,9 +27,14 @@ def query(interval)
         ]]
       end
 
+      # :reek:DuplicateMethodCall
```
Comment (santib) [inline app/services/metrics/defect_escape_rate/per_project.rb:30]: I'd prefer to have these comments in `.reek.yml` rather than polluting the actual code with them

File: app/services/metrics/defect_escape_rate/per_project.rb (hunk 1 above)
Comment (santib) [reply app/services/metrics/defect_escape_rate/per_project.rb:32]: Agree with Hora, but actually, I think it should start in 0 by default and always add 1

```ruby
bug_issues.each_with_object(Hash.new(0)) do |bug, result|
  result[bug.environment] += 1
end
```

File: app/views/development_metrics/defect_escape_rate/_values_details.erb [hunk 4]
```diff
@@ -18,7 +18,11 @@
         <h5>Bugs by environment:</h5>
         <% metric[:values].each do |metric| %>
           <div class="item-list">
+            <% if metric.first -%>
             <%= metric.first.to_s.capitalize %>:
+            <% else -%>
+              None:
+            <% end -%>
```
Comment (santib) [inline app/views/development_metrics/defect_escape_rate/_values_details.erb:25]:
```suggestion
            <%= metric.first.to_s.capitalize.presence || 'None:' %>:
```

Apart from that, I don't remember if we have presenters in this project, if we do have it might go there.

File: app/views/development_metrics/defect_escape_rate/_values_details.erb [hunk 5]
```diff
@@ -18,7 +18,11 @@
         <h5>Bugs by environment:</h5>
         <% metric[:values].each do |metric| %>
           <div class="item-list">
+            <% if metric.first -%>
             <%= metric.first.to_s.capitalize %>:
+            <% else -%>
+              None:
```
Comment (santib) [inline app/views/development_metrics/defect_escape_rate/_values_details.erb:24]: we are missing a test checking for `None:`

File: app/views/development_metrics/defect_escape_rate/_values_details.erb [hunk 6]
```diff
@@ -18,7 +18,11 @@
         <h5>Bugs by environment:</h5>
         <% metric[:values].each do |metric| %>
```
Comment (santiagovidal) [inline app/views/development_metrics/defect_escape_rate/_values_details.erb:19]: Instead of doing `metric.first` and `metric.second`, we can define this loop like:

`metric[:values].each do |environment, bugs|`

It's easier to follow

File: app/services/metrics/defect_escape_rate/per_project.rb (hunk 1 above)
Comment (enzofab91) [reply app/services/metrics/defect_escape_rate/per_project.rb:32]: Nice!

File: app/views/development_metrics/defect_escape_rate/_values_details.erb (hunk 4 above)
Comment (enzofab91) [reply app/views/development_metrics/defect_escape_rate/_values_details.erb:25]: We don't have presenters in this project yet.

File: Gemfile.lock (hunk 2 above)
Comment (enzofab91) [reply Gemfile.lock:152]: When I did `bundle install` I have to update this gem in order to work. You think I have to remove this file from PR?

File: app/views/development_metrics/defect_escape_rate/_values_details.erb (hunk 6 above)
Comment (enzofab91) [reply app/views/development_metrics/defect_escape_rate/_values_details.erb:19]: Nice, done!

File: app/views/development_metrics/defect_escape_rate/_values_details.erb (hunk 5 above)
Comment (enzofab91) [reply app/views/development_metrics/defect_escape_rate/_values_details.erb:24]: Done!

File: .rspec [hunk 7]
```diff
@@ -1 +1 @@
---require spec_helper
+--require rails_helper
```
Comment (santib) [inline .rspec:1]: I think this was fine, we should require `spec_helper` here. Why was the change?

File: Gemfile.lock (hunk 2 above)
Comment (santib) [reply Gemfile.lock:152]: we can leave it this time

File: app/views/development_metrics/defect_escape_rate/_values_details.erb [hunk 8]
```diff
@@ -16,10 +16,10 @@
       <hr>
       <div>
         <h5>Bugs by environment:</h5>
-        <% metric[:values].each do |metric| %>
+        <% metric[:values].each do |environment, bugs| %>
           <div class="item-list">
-            <%= metric.first.to_s.capitalize %>:
-              <%= metric.second.to_s %>
+            <%= environment.to_s.capitalize.presence || 'None' %>:
+            <%= bugs.second.to_s %>
```
Comment (santib) [inline app/views/development_metrics/defect_escape_rate/_values_details.erb:22]: shouldn't this be
```suggestion
            <%= bugs %>
```

now that you changed `|metric|` for `|environment, bugs|` ?

File: spec/controllers/development_metrics_controller_spec.rb [hunk 9]
```diff
@@ -1,11 +1,12 @@
-require 'rails_helper'
```
Comment (santib) [inline spec/controllers/development_metrics_controller_spec.rb:1]: we should keep this line here

File: .rspec (hunk 7 above)
Comment (enzofab91) [reply .rspec:1]: `rails_helper` includes `spec_helper`, and allows me to remove `require 'rails_helper'` from `spec` files.

File: spec/controllers/development_metrics_controller_spec.rb (hunk 9 above)
Comment (enzofab91) [reply spec/controllers/development_metrics_controller_spec.rb:1]: Related to the comment above.

File: spec/controllers/development_metrics_controller_spec.rb [hunk 10]
```diff
@@ -44,10 +45,82 @@
...
+          before do
+            subject
+          end
+
+          it 'has a successful response' do
+            expect(response).to be_successful
+          end
+
+          it 'render EDR metric with correct production issues' do
+            expect(response.body).to include('Production:' + NEWLINE_WHITESPACES +
```
Comment (santib) [inline spec/controllers/development_metrics_controller_spec.rb:107]: the `NEWLINE_WHITESPACES` thing looks fragile. Why not doing `/Production:\s*#{production_jira_bugs.count}/` ?

Also don't use `+` for string interpolation, instead use `"#{} #{}`

File: app/views/development_metrics/defect_escape_rate/_values_details.erb (hunk 8 above)
Comment (enzofab91) [reply app/views/development_metrics/defect_escape_rate/_values_details.erb:22]: Haha yes 😅

Comment (santib) [inline app/views/development_metrics/defect_escape_rate/_values_details.erb:22]:
```suggestion
            <%= environment.to_s.capitalize.presence || 'None' %>: <%= bugs.second.to_s %>
```

is there a reason to not have it like that?

File: .rspec (hunk 7 above)
Comment (santib) [reply .rspec:1]: https://github.com/rspec/rspec-rails/issues/1180

File: app/views/development_metrics/defect_escape_rate/_values_details.erb (hunk 8 above)
Comment (enzofab91) [reply app/views/development_metrics/defect_escape_rate/_values_details.erb:22]: No reason why, I'll change it.

**Review by santib — APPROVED** [review]
good job 💪

**Review by santiagovidal — APPROVED** [review]

**Review by horacio — APPROVED** [review]
